daffodil-io/src/main/scala/org/apache/daffodil/io/Dump.scala

```diff
@@ -577,18 +574,20 @@ class DataDumper {
         bb.flip()
 
         Assert.invariant(bb.remaining > 0)
-        do {
-          // An overflow means we were able to start to decode at least 1 sequence of characters, but there was either insufficient
-          // space in the output buffer to store said decoded char or there were left over bytes after parsing. If it is
-          // the former, we can proceed and we'll get the left over bytes on the next run, if it was the latter
-          // (as can be the case with decoding a 4 byte character sequence), we will call decode with a larger buffer
-          // until we consume something or the output buffer is at same capacity as input buffer
-          cr = dec.decode(bb, cb, true)
-          nConsumedBytes = bb.position()
-          if (cr.isOverflow && nConsumedBytes == 0) {
-            cb = CharBuffer.allocate(cb.capacity + 1)
-          }
-        } while (cr.isOverflow && nConsumedBytes == 0 && cb.capacity <= bb.capacity)
+        while ({
+          {
+            // An overflow means we were able to start to decode at least 1 sequence of characters, but there was either insufficient
+            // space in the output buffer to store said decoded char or there were left over bytes after parsing. If it is
+            // the former, we can proceed and we'll get the left over bytes on the next run, if it was the latter
+            // (as can be the case with decoding a 4 byte character sequence), we will call decode with a larger buffer
+            // until we consume something or the output buffer is at same capacity as input buffer
+            cr = dec.decode(bb, cb, true)
+            nConsumedBytes = bb.position()
+            if (cr.isOverflow && nConsumedBytes == 0) {
+              cb = CharBuffer.allocate(cb.capacity + 1)
+            }
+          }; cr.isOverflow && nConsumedBytes == 0 && cb.capacity <= bb.capacity
+        }) ()
```

Comment:
This is pretty hard to follow. I *think* the loop condition still evaluates to true on the first iteration, so can this be converted to a normal while loop by just moving the condition up, e.g.

```scala
while (cr.isOverflow && nConsumedBytes == 0 && cb.capacity <= bb.capacity) {
  // normal loop content
  ...
}
```
It means the first time entering the loop will always evaluate to true, but avoiding that with a do-while isn't that super performance critical--this section of code is never in a hot loop.